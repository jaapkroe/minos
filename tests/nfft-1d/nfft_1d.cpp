/*
 * Copyright (c) 2002, 2016 Jens Keiner, Stefan Kunis, Daniel Potts
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>

#define NFFT_PRECISION_DOUBLE

#include "nfft3mp.h"

static void simple_test_nfft_1d(void)
{
  NFFT(plan) p;

  int N = 14; // number of fourier coefficients
  int M = 24; // number of data points
  int i;
  double L;
  FILE* f;
  char c[16];
  const char *error_str;

  // read data file
  f = fopen("nfft_1d.dat","r");
  fscanf(f, "%s %d %d %lf", c, &M, &N, &L);                                   // number of points, coefficients and box size
  NFFT(init_1d)(&p, N, M);                                                    // initialize the FFT plan (p) and allocate arrays for p
  for(i=0;i<M;i++) fscanf(f, "%lf %lf %lf", &p.x[i], &p.f[i][0], &p.f[i][1]); // read data file
  fclose(f);
  for(i=0;i<M;i++) { p.x[i]/=L; p.x[i] -= round(p.x[i]); }                    // wrap data to cell [-0.5, 0.5)
  f=fopen("input-check.dat","w"); for(i=0;i<M;i++) fprintf(f,"%-3d %12.8g %12.8g %12.8g\n",i,p.x[i],p.f[i][0],p.f[i][1]); fclose(f);

  //  /** init pseudo random nodes */
  NFFT(vrand_shifted_unit_double)(p.x, p.M_total);

  /** precompute psi, the entries of the matrix B */
  if (p.flags & PRE_ONE_PSI) NFFT(precompute_one_psi)(&p);

  /** init pseudo random Fourier coefficients and show them */
  NFFT(vrand_unit_complex)(p.f_hat,p.N_total);
  NFFT(vpr_complex)(p.f_hat, p.N_total, "given Fourier coefficients, vector f_hat");

  /** check for valid parameters before calling any trafo/adjoint method */
  error_str = NFFT(check)(&p);
  if (error_str != 0)
  {
    printf("Error in nfft module: %s\n", error_str);
    return;
  }

  /** direct trafo and show the result */
  NFFT(trafo_direct)(&p);
  NFFT(vpr_complex)(p.f,p.M_total,"ndft, vector f");
  f=fopen("output_direct.dat","w"); for(i=0;i<M;i++) fprintf(f,"%-3d %12.8g %12.8g %12.8g\n",i,p.x[i],p.f[i][0],p.f[i][1]); fclose(f);

  // /** approx. trafo and show the result */
  NFFT(trafo)(&p);
  NFFT(vpr_complex)(p.f,p.M_total,"nfft, vector f");
  f=fopen("output_approx.dat","w"); for(i=0;i<M;i++) fprintf(f,"%-3d %12.8g %12.8g %12.8g\n",i,p.x[i],p.f[i][0],p.f[i][1]); fclose(f);

  // /** approx. adjoint and show the result */
  NFFT(adjoint_direct)(&p);
  NFFT(vpr_complex)(p.f_hat,p.N_total,"adjoint ndft, vector f_hat");
  f=fopen("output_adj_direct.dat","w"); for(i=0;i<N;i++) fprintf(f,"%-3d %12.8g %12.8g\n",i,p.f_hat[i][0],p.f_hat[i][1]); fclose(f);

  // /** approx. adjoint and show the result */
  NFFT(adjoint)(&p);
  NFFT(vpr_complex)(p.f_hat,p.N_total,"adjoint nfft, vector f_hat");
  f=fopen("output_adj_approx.dat","w"); for(i=0;i<N;i++) fprintf(f,"%-3d %12.8g %12.8g\n",i,p.f_hat[i][0],p.f_hat[i][1]); fclose(f);

  /** finalise the one dimensional plan */
  NFFT(finalize)(&p);
}

int main(void)
{
  printf("1) computing a one dimensional ndft, nfft and an adjoint nfft\n\n");
  simple_test_nfft_1d();

  return EXIT_SUCCESS;
}
