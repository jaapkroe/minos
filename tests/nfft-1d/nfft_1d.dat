#  64 256 1
         0            0            0
   0.01562     0.980171            0
   0.03125       1.9509            0
   0.04688      2.90285            0
    0.0625      3.82683            0
   0.07812      4.71397            0
   0.09375       5.5557            0
    0.1094      6.34393            0
     0.125      7.07107            0
    0.1406       7.7301            0
    0.1562       8.3147            0
    0.1719      8.81921            0
    0.1875       9.2388            0
    0.2031       9.5694            0
    0.2188      9.80785            0
    0.2344      9.95185            0
      0.25           10            0
    0.2656      9.95185            0
    0.2812      9.80785            0
    0.2969       9.5694            0
    0.3125       9.2388            0
    0.3281      8.81921            0
    0.3438       8.3147            0
    0.3594       7.7301            0
     0.375      7.07107            0
    0.3906      6.34393            0
    0.4062       5.5557            0
    0.4219      4.71397            0
    0.4375      3.82683            0
    0.4531      2.90285            0
    0.4688       1.9509            0
    0.4844     0.980171            0
       0.5  1.22465e-15            0
    0.5156    -0.980171            0
    0.5312      -1.9509            0
    0.5469     -2.90285            0
    0.5625     -3.82683            0
    0.5781     -4.71397            0
    0.5938      -5.5557            0
    0.6094     -6.34393            0
     0.625     -7.07107            0
    0.6406      -7.7301            0
    0.6562      -8.3147            0
    0.6719     -8.81921            0
    0.6875      -9.2388            0
    0.7031      -9.5694            0
    0.7188     -9.80785            0
    0.7344     -9.95185            0
      0.75          -10            0
    0.7656     -9.95185            0
    0.7812     -9.80785            0
    0.7969      -9.5694            0
    0.8125      -9.2388            0
    0.8281     -8.81921            0
    0.8438      -8.3147            0
    0.8594      -7.7301            0
     0.875     -7.07107            0
    0.8906     -6.34393            0
    0.9062      -5.5557            0
    0.9219     -4.71397            0
    0.9375     -3.82683            0
    0.9531     -2.90285            0
    0.9688      -1.9509            0
    0.9844    -0.980171            0
