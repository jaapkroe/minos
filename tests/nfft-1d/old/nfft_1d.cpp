#include <iostream>
#include <fstream>
#include <fftw3.h>
#include <nfft3.h>
#define NMAX 8192
int main()
{
  int i, N, M=16;
  double L;
  char tmp[16];
  std::ifstream file("nfft_1d.dat");
  file >> tmp >> N >> L;
  double *input = new double[N];
  double *grid = new double[N];
  if(N>NMAX) { fprintf(stderr, "N>NMAX (%d>%d)\n",N,NMAX); exit(1); };
  for(i=0;i<N;i++) file >> grid[i] >> input[i];

  nfft_plan p;
  fprintf(stdout,"init_1d:\n");
  nfft_init_1d(&p,N,M);
  if(p.flags & PRE_ONE_PSI) nfft_precompute_one_psi(&p);
  fprintf(stdout,"vrand shifted unit double:\n");
  nfft_vrand_shifted_unit_double(p.x,p.M_total);
  fprintf(stdout,"vpr_complex:\n");
  nfft_vpr_complex(p.f_hat,p.N_total,"given Fourier coefficients, f_hat");

  fprintf(stdout,"trafo:\n");
  nfft_trafo(&p);

  fprintf(stdout,"finalize:\n");
  nfft_finalize(&p);

  //if(M>N) M=N;
  fprintf(stdout,"read %d data points, generating FFT with %d q-points\n",N,M);

  //fftw_complex *out = fftw_alloc_complex(M);
  //fftw_plan p = fftw_plan_dft_r2c_1d(N, input, out, FFTW_ESTIMATE);
  //fftw_execute(p);

  //for(i=0;i<M;i++) printf("%-4d %12.4g %12.4g\n",i,out[i][0]/N,out[i][1]/N);
}
