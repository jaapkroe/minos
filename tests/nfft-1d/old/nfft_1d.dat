#  64 1
      0.0000       0.00000373
      0.0982       0.00000804
      0.1963       0.00001693
      0.2945       0.00003479
      0.3927       0.00006977
      0.4909       0.00013654
      0.5890       0.00026075
      0.6872       0.00048595
      0.7854       0.00088383
      0.8836       0.00156868
      0.9817       0.00271706
      1.0799       0.00459264
      1.1781       0.00757568
      1.2763       0.01219489
      1.3744       0.01915717
      1.4726       0.02936852
      1.5708       0.04393693
      1.6690       0.06414674
      1.7671       0.09139376
      1.8653       0.12707367
      1.9635       0.17242162
      2.0617       0.22831003
      2.1598       0.29502266
      2.2580       0.37203419
      2.3562       0.45783336
      2.4544       0.54983088
      2.5525       0.64438872
      2.6507       0.73699381
      2.7489       0.82257756
      2.8471       0.89595662
      2.9452       0.95234480
      3.0434       0.98786717
      3.1416       1.00000000
      3.2398       0.98786717
      3.3379       0.95234480
      3.4361       0.89595662
      3.5343       0.82257756
      3.6325       0.73699381
      3.7306       0.64438872
      3.8288       0.54983088
      3.9270       0.45783336
      4.0252       0.37203419
      4.1233       0.29502266
      4.2215       0.22831003
      4.3197       0.17242162
      4.4179       0.12707367
      4.5160       0.09139376
      4.6142       0.06414674
      4.7124       0.04393693
      4.8106       0.02936852
      4.9087       0.01915717
      5.0069       0.01219489
      5.1051       0.00757568
      5.2033       0.00459264
      5.3014       0.00271706
      5.3996       0.00156868
      5.4978       0.00088383
      5.5960       0.00048595
      5.6941       0.00026075
      5.7923       0.00013654
      5.8905       0.00006977
      5.9887       0.00003479
      6.0868       0.00001693
      6.1850       0.00000804
