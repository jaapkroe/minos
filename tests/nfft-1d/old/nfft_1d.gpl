#!/usr/bin/gnuplot

set term pdfcairo enh size 4,2 lw 1.5 dashed
set output 'nfft_1d.pdf'

set xran [0:]
set xlab "n"
set ylab "f(n)"

N=system("grep N= nfft_1d-data_generatation.sh | sed 's/.*=//g'")+0
pl 'nfft_1d.dat' u 1 w p ls 1 ps .5 not,\
   'nfft_1d.gen' u ($1/(2*pi)*N):2 w l ls 1 dt 1 not,\

set boxwidth 0.2
set style fill solid 

set key outside above
set xlab "q"
set ylab "a_q"
set lab 1 "f(x) = a_0 + 2{/Symbol S}_{q>0}a_qe^{-iqx}" right at graph .9,.85

pl 'nfft_1d.out' u ($1+0.0):2 w boxes ls 2 t 'Re(z)',\
   'nfft_1d.out' u ($1+0.2):3 w boxes ls 4 t 'Im(z)',\
   'nfft_1d.out' u ($1+0.4):(sqrt($2**2+$3**2)) w boxes ls 1 lw 2 t "|z|"

