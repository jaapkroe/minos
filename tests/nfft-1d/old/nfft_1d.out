vpr_complex:

 given Fourier coefficients, f_hat, adr=0x15f65a0
     0.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
     4.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
     8.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    12.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    16.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    20.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    24.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    28.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    32.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    36.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    40.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    44.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    48.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    52.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    56.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
    60.	+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,+0.0E+00+0.0E+00i,
trafo:
finalize:
read 64 data points, generating FFT with 16 q-points
